### ISSUE TITLE
FB not recognized

### ISSUE THREAD

**Issue by mdecesar** [nanograv/PINT#263]
PINT doesn't recognize the binary frequency, FB, or FB derivatives.  (Guessing this may already be a known issue.)

Example:  I ran PINT on the J0023+0923 parfile, which has binary parameters FB0 through FB3.  I got a warning saying these FB parameters are not recognized by PINT, and when I write the parameters as a parfile, the FB params are not included.

WARNING: Unrecognized parfile line 'FB0     8.338721549327D-05  1  2.075175747949D-14' [pint.models.timing_model]
WARNING: Unrecognized parfile line 'FB1    -1.884380737910D-20  1  1.399414153825D-21' [pint.models.timing_model]
WARNING: Unrecognized parfile line 'FB2     2.404732960981D-28  1  4.634445746334D-29' [pint.models.timing_model]
WARNING: Unrecognized parfile line 'FB3     2.286833746017D-35  1  4.318133111493D-36' [pint.models.timing_model]

**Comment by paulray** [comment]
So far, PINT has no support for FB parameters. This is somewhat related to issue #111.

There are two ways to add FB support.  A variant of each model could be defined, like BT and BTX, where BTX uses the FB parameterization and BT uses PB, PBDOT.   In my view this is not optimal.  I'd rather see a modular solution where there the calculation of the orbital phase is modular and can be used with any of the orbital delay models (ELL1, BT, DD, etc...). Then it would be easy to use (PB, PBDOT), (FB0, FB1, FB2, ...) or a more complex model that maybe had sinusoidal modulations of the orbital phase on top of the basic periodicity.  This is useful for systems like redbacks that have a lot of orbital period noise, in the same way that WAVE parameters are useful for pulsars with lots of spin noise.   But, that may be asking too much.  For sure the quick solution is for someone to just add a BTX model to PINT...

**Comment by scottransom** [comment]
I completely agree with your recommendation, Paul.  I think that modular approach would be *much* better.  And we'd only need one Kepler's Eqn solver then as well.

**Comment by paulray** [comment]
What I don't know is whether the models we have implemented are easily separable like that. Are the delay calculations clearly separable from the phase calculation in them?  I haven't looked at the guts.

**Comment by luojing1211** [comment]
@paulray The way it is right now is we have the orbit phase calculation ( Use PB, PBDOT) in a generic binary model. All the specific binary model is an inheritance of the generic one. If any model is using some other parameterization like ELL1, it overrides the generic binary orbit phase calculation. Some in a sense, it has the potential to set it up as you recommended.

**Comment by luojing1211** [comment]
Is there any BTX paper I can reference to implement this?

**Comment by paulray** [comment]
I don't know of any BTX paper, but I think it is nothing more than the difference between P0, P1 and F0, F1, F2, .. for spin phase calculations, but for orbital phase.

But, I think this is a place where we might benefit from some thinking about the design, before jumping in to try to patch in a fix. (Unless the fix is quite easy and gets FBx going quickly while we consider the larger issues)

I've been thinking about modularizing the orbital phase and delay calculations, and other refactoring of the timing model (e.g. #242), and there are still some things I'm not sure about.  For example, if they are modularized, how would models like DDGR be handled, where the orbital phase evolution (e.g. OMDOT) is *determined* by the other parameters in the delay model? Is it really sensible to modularize the phase and delay calculations?  And, how would the structure generalize in the future to handle more complex systems (e.g. multi-planet or triple systems). We don't need to implement those right away, but we should have an idea of a path to that.

A good way to start might be to make a UML diagram of the current situation, and one for a proposed solution.

**Comment by luojing1211** [comment]
This issue is fix by the PR #356, should we close it?

**Comment by paulray** [comment]
Maybe @mdecesar should test whether it works now on her J0023 par file?
